src/platform/macos_iokit/enumeration.rs

```diff
@@ -147,6 +157,73 @@ fn get_children(device: &IoService) -> Result<IoServiceIterator, Error> {
     }
 }
 
+fn get_parent(device: &IoService) -> Result<IoService, Error> {
+    unsafe {
+        let mut handle = 0;
+        let r = IORegistryEntryGetParentEntry(device.get(), kIOServicePlane as *mut _, &mut handle);
+        if r != kIOReturnSuccess {
+            debug!("IORegistryEntryGetParentEntry failed: {r}");
+            return Err(Error::from_raw_os_error(r));
+        }
+
+        Ok(IoService::new(handle))
+    }
+}
+
+fn get_port_number(device: &IoService) -> Option<u32> {
+    get_integer_property::<u32>(device, "PortNum").or_else(|| {
+        if let Ok(parent) = get_parent(device) {
+            return get_data_property(&parent, "port")
+                .map(|d| u32::from_ne_bytes(d[0..4].try_into().unwrap()));
+        }
+        None
+    })
+}
+
+fn get_port_chain(device: &IoService) -> impl Iterator<Item = u32> {
+    let mut port_chain = VecDeque::new();
+
+    if let Some(port_number) = get_port_number(device) {
+        port_chain.push_back(port_number);
+    }
+
+    if let Ok(mut hub) = get_parent(device) {
+        loop {
+            let port_number = match get_port_number(&hub) {
+                Some(p) => p,
+                None => break,
+            };
+            if port_number == 0 {
+                break;
+            }
+            port_chain.push_front(port_number);
+
+            let session_id = match get_integer_property::<u64>(&hub, "sessionID") {
+                Some(session_id) => session_id,
+                None => break,
+            };
+
+            hub = match get_parent(&hub) {
+                Ok(hub) => hub,
+                Err(_) => break,
+            };
+
+            // Ignore the same sessionID
+            if session_id
```

Comment:
Did you observe it reporting a device as its own parent or find documentation indicating that, or is this just a precaution to prevent infinite-looping?

Reply:
The parent device may return the same `sessionID`, which needs to be ignored until a new `sessionID` is obtained or the obtaining fails.